### PR TITLE
RMF security section

### PR DESCRIPTION
**Pull request by marcoag**

### CODE REVIEW

File: src/security.md [hunk 1]
```diff
@@ -1,4 +1,279 @@
...
 This chapter describes how to use DDS Security tools to provide authentication,
-encryption, and access control to a RoMi-H system.
+encryption, and access control to a RoMi-H system. 
+
+The security of the RoMi-H system can be divided into two main parts: 
+its ROS 2 elements and the dashboard. The security of the ROS 2 
+elements is provided by the DDS security tools which help ensure authentication, 
+encryption and access control. The dashboard provides the user with an 
+instruments panel while ensuring encryption, integrity and authentication of 
+the server side through TLS. The user authentication and access control
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
the connection to the server through TLS. User authentication and access control 
```

File: src/security.md [hunk 2]
```diff
@@ -1,4 +1,279 @@
...
+encryption, and access control to a RoMi-H system. 
+
+The security of the RoMi-H system can be divided into two main parts: 
+its ROS 2 elements and the dashboard. The security of the ROS 2 
+elements is provided by the DDS security tools which help ensure authentication, 
+encryption and access control. The dashboard provides the user with an 
+instruments panel while ensuring encryption, integrity and authentication of 
+the server side through TLS. The user authentication and access control 
+is made by user/password checking against a database and then 
+providing that user with an access level to the secured ROS 2 network that
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
 providing that user with access to the secured ROS 2 network at a level that
```

File: src/security.md [hunk 3]
```diff
@@ -1,4 +1,279 @@
...
+corresponds to the role of that user.
+ 
+![Security System Infrastructure](images/security/system_infrastructure.png)
+
+As a guidance, the [RMF demos][rmf_demos] repository contains an example of a
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
The [RMF demos][rmf_demos] repository contains an example of a 
```

File: src/security.md [hunk 4]
```diff
@@ -1,4 +1,279 @@
...
+instruments panel while ensuring encryption, integrity and authentication of 
+the server side through TLS. The user authentication and access control 
+is made by user/password checking against a database and then 
+providing that user with an access level to the secured ROS 2 network that
+corresponds to the role of that user.
+ 
+![Security System Infrastructure](images/security/system_infrastructure.png)
+
+As a guidance, the [RMF demos][rmf_demos] repository contains an example of a 
+full RMF run of the office demo using secured ROS 2 communications along with
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
full RMF application using secured ROS 2 communications along with
```

File: src/security.md [hunk 5]
```diff
@@ -1,4 +1,279 @@
...
+step by step explanations. 
+
+## ROS 2 Security
+
+ROS 2 contains tools that help create and load the needed security artifacts 
+to enable DDS-security. RoMi-H makes uses of these tools in order to enable 
+security on its ROS 2 elements. A brief introduction to these tools and its 
+usage is provided here, however, for a deeper understanding of the whole system,
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
usage is provided here. For a deeper understanding of the whole system, 
```

File: src/security.md [hunk 6]
```diff
@@ -1,4 +1,279 @@
...
+full RMF run of the office demo using secured ROS 2 communications along with
+step by step explanations. 
+
+## ROS 2 Security
+
+ROS 2 contains tools that help create and load the needed security artifacts 
+to enable DDS-security. RoMi-H makes uses of these tools in order to enable 
+security on its ROS 2 elements. A brief introduction to these tools and its 
+usage is provided here, however, for a deeper understanding of the whole system, 
+please refer to the documenation of the [ROS 2 DDS-Security integration][SROS].
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
please refer to the [ROS 2 DDS-Security integration][SROS] documentation.
```

File: src/security.md [hunk 7]
```diff
@@ -1,4 +1,279 @@
...
+just run the participant in non secure mode if those files are not found. 
+- **ROS_SECURITY_KEYSTORE** should point to the root of the keystore directory tree. This will help `RCL`
+find the location of the security artifacts to initialize the ROS 2 security environment.
+
+#### SROS 2 security keystore
+
+The keystore is the root directory where the DDS security artifacts are stored. `RCL` will use the 
+contents of this directory to provide the DDS security to the ROS 2 network. The 
+`ROS_SECURITY_KEYSTORE` environment variable should by convention point to this directory. In order 
+to initalize and populate a `keystore_storage` directory files the following command can be used:
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
to initalize and populate `keystore_storage` directory files the following command can be used:
```

File: src/security.md [hunk 8]
```diff
@@ -1,4 +1,279 @@
...
+    |   ├── identity_ca.key.pem
+    |   └── permissions_ca.key.pem
+    └── public
+        ├── ca.key.pem
+        ├── identity_ca.key.pem
+        └── permissions_ca.key.pem
+
+The `public` directory contains anything permissible as public, such as public certificates for 
+the identity or permissions Certificate Authorities (CA). As such, this can be given read access to all 
+executables. Note that in the default case, both the `identity_ca` and `permissions_ca` points to the
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
executables. Note that in the default case, both `identity_ca` and `permissions_ca` point to the 
```

File: src/security.md [hunk 9]
```diff
@@ -1,4 +1,279 @@
...
+        ├── identity_ca.key.pem
+        └── permissions_ca.key.pem
+
+The `public` directory contains anything permissible as public, such as public certificates for 
+the identity or permissions Certificate Authorities (CA). As such, this can be given read access to all 
+executables. Note that in the default case, both the `identity_ca` and `permissions_ca` points to the 
+same CA certificate.
+
+The `private` directory contains anything permissible as private, such as private key material for 
+aforementioned certificate authorities. This directory should be redacted before deploying the
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
aforementioned certificate authorities. This directory should be removed before deploying the 
```

File: src/security.md [hunk 10]
```diff
@@ -1,4 +1,279 @@
...
+
+The `private` directory contains anything permissible as private, such as private key material for 
+aforementioned certificate authorities. This directory should be redacted before deploying the 
+keystore onto the target device/robot.
+
+The `enclaves` directory contains the security artifacts associated with individual enclaves. 
+This is a concept that was introduced with the `ROS 2 Foxy Fitzroy` release, as domain participants 
+are no longer mapped directly to ROS nodes. `Foxy` introduces the concept of a security “enclave”, 
+where an “enclave” is a process or group of processes that will share the same identity and access 
+control rules. The enclaves folder may still recursively nest sub-paths for organizing separate enclaves.
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
control rules. The enclaves folder may recursively nest sub-paths for organizing separate enclaves. 
```

File: src/security.md [hunk 11]
```diff
@@ -1,4 +1,279 @@
...
+same CA certificate.
+
+The `private` directory contains anything permissible as private, such as private key material for 
+aforementioned certificate authorities. This directory should be redacted before deploying the 
+keystore onto the target device/robot.
+
+The `enclaves` directory contains the security artifacts associated with individual enclaves. 
+This is a concept that was introduced with the `ROS 2 Foxy Fitzroy` release, as domain participants 
+are no longer mapped directly to ROS nodes. `Foxy` introduces the concept of a security “enclave”, 
+where an “enclave” is a process or group of processes that will share the same identity and access
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
The `enclaves` directory contains the security artifacts associated with individual security enclaves. 
SROS 2 introduces the concept of a security “enclave”, 
where an “enclave” is a process or group of processes that will share the same identity and access 
```

File: src/security.md [hunk 12]
```diff
@@ -1,4 +1,279 @@
...
+
+The `enclaves` directory contains the security artifacts associated with individual enclaves. 
+This is a concept that was introduced with the `ROS 2 Foxy Fitzroy` release, as domain participants 
+are no longer mapped directly to ROS nodes. `Foxy` introduces the concept of a security “enclave”, 
+where an “enclave” is a process or group of processes that will share the same identity and access 
+control rules. The enclaves folder may still recursively nest sub-paths for organizing separate enclaves. 
+
+#### SROS 2 enclave keys
+
+Once the `keystore` has been initialized you might wanna create the security keys for your
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
Once the `keystore` has been initialized you may wish to create the security keys for your 
```

File: src/security.md [hunk 13]
```diff
@@ -1,4 +1,279 @@
...
+/hospital/fleet_state_visualizer
+/hosptial/loop_request_publisher
+/hosptial/rviz2
+...
+```
+
+## RoMi-H Dashboard Security
+
+The RoMi-H Dashboard is a web application that provides overall visualization and control over the 
+RoMi-H system. It is serverd over TLS to ensure encryption, integrity and authentication of the
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
RoMi-H system. It is served over TLS to ensure encryption, integrity and authentication of the 
```

File: src/security.md [hunk 14]
```diff
@@ -1,4 +1,279 @@
...
+/hosptial/rviz2
+...
+```
+
+## RoMi-H Dashboard Security
+
+The RoMi-H Dashboard is a web application that provides overall visualization and control over the 
+RoMi-H system. It is serverd over TLS to ensure encryption, integrity and authentication of the 
+communication with the final user. The server uses [openid-connect (OIDC)][openid_connect] for 
+authentication, it is a open standard based on [oauth 2.0][oauth2] and [JOSE][jose]. Currently the
```
Comment (gbiggs) [inline src/security.md]:
```suggestion
authentication, an open standard based on [oauth 2.0][oauth2] and [JOSE][jose]. Currently the 
```

**Review by gbiggs — COMMENTED** [review]
Not too many changes, but I did not do detailed corrections for grammar. I can do that if you wish.

Overall, this chapter feels like it needs a more detailed example that it can walk through step by step.

**Comment by marcoag** [comment]
> Not too many changes, but I did not do detailed corrections for grammar. I can do that if you wish.

However you wish, don't wanna give you extra work :D

> Overall, this chapter feels like it needs a more detailed example that it can walk through step by step.

Totally agree, I thought of doing a tutorial but then I thought it might be better to have a tutorial section (including tutorials for other features of RMF) and then just add a link here to it, what are your thoughts on that?

**Comment by gbiggs** [comment]
If it's possible to include files from other repositories automatically somehow, then doing a tutorial elsewhere and linking it in as content would be awesome. If not then I agree with doing a tutorial in its own repository somewhere and just linking it.

**Comment by marcoag** [comment]
Perfect, I will look into a place for making the tutorials. I'll merge this one in the meantime and add the extra info later.

**Pull request merged**